src/Namespace.sol

```diff
@@ -0,0 +1,578 @@
+
+import {Owned} from "../lib/solmate/src/auth/Owned.sol";
+
+import {FixedPointMathLib} from "../lib/solmate/src/utils/FixedPointMathLib.sol";
+
+error InsufficientFunds(); // The transaction does not have enough money to pay for this.
+error Unauthorized(); // The caller is not authorized to perform this action.
+
+error InvalidCommit(); // The commitment hash was not found
+error InvalidName(); // The username had invalid characters
+error InvalidTime(); // Time is too far in the future or past
+error IncorrectOwner(); // The username is not owned by the expected address
+
+error Registered(); // The username is currently registered.
+error NotRegisterable(); // The username has been registered and cannot be registered again.
+error Registerable(); // The username has never been registered.
+
+error Expired(); // The username is expired (renewable or biddable)
+error Biddable(); // The username is biddable
+error NotBiddable(); // The username is still registered or in renewal.
+
+error Escrow(); // The recovery request is still in escrow
+error NoRecovery(); // The recovery request could not be found
+error InvalidRecovery(); // The recovery address is being set to the custody address
+
+/**
+ * @title Namespace
+ * @author varunsrin
+ */
+contract Namespace is ERC721, Owned {
+    using FixedPointMathLib for uint256;
+
+    /*//////////////////////////////////////////////////////////////
+                        REGISTRATION EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    event Renew(uint256 indexed tokenId, address indexed to, uint256 expiry);
+
+    /*//////////////////////////////////////////////////////////////
+                        RECOVERY EVENTS
+    //////////////////////////////////////////////////////////////*/
+
+    event SetRecoveryAddress(address indexed recovery, uint256 indexed tokenId);
+
+    event RequestRecovery(uint256 indexed id, address indexed from, address indexed to);
+
+    event CancelRecovery(uint256 indexed id);
+
+    /*//////////////////////////////////////////////////////////////
+                        REGISTRATION STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    // Mapping from commitment hash to state
+    mapping(bytes32 => bool) public stateOf;
+
+    // Mapping from tokenID to expiration year
+    mapping(uint256 => uint256) public expiryOf;
+
+    // The index of the next year in the array
+    uint256 internal _nextYearIdx;
+
+    /*//////////////////////////////////////////////////////////////
+                        RECOVERY STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    // Mapping from tokenId to recoveryAddress
+    mapping(uint256 => address) public recoveryOf;
+
+    // Mapping from tokenId to recovery timestamp in seconds, which is set to zero on cancellation
+    // or completion
+    mapping(uint256 => uint256) public recoveryClockOf;
```

Comment:
I'm not a gas golf expert, but curious if having a mapping of `(uint256 => recoveryObj)` where recoveryObj is a struct with recovery address, clock and recoveryDestination would be more efficient.